cg/services/order_validation_service/workflows/tomte/validation/inter_field/rules.py

```diff
@@ -31,9 +30,35 @@
 
 
 def validate_wells_contain_at_most_one_sample(order: TomteOrder) -> list[OccupiedWellError]:
-    samples_with_cases = _get_plate_samples(order)
-    samples = _get_excess_samples(samples_with_cases)
-    return _get_errors(samples)
+    errors = []
+    well_position_to_sample_map = create_well_position_to_sample_map(order)
+    for indices in well_position_to_sample_map.values():
+        if len(indices) > 1:
+            well_errors = get_occupied_well_errors(indices[1:])
+            errors.extend(well_errors)
+    return errors
+
+
+def create_well_position_to_sample_map(
+    order: TomteOrder,
+) -> dict[tuple[str, str], list[tuple[int, int]]]:
+    """
+    Constructs a dict with keys being a (container_name, well_position) pair. For each such pair, the value will be
+    a list of (case index, sample index) pairs corresponding to all samples with matching container_name and
+    well_position, provided the sample is on a plate.
+    """
+    well_position_to_sample_map = {}
+    for case_index, case in order.enumerated_cases:
+        for sample_index, sample in case.enumerated_samples:
+            if _is_sample_on_plate(sample):
+                if not well_position_to_sample_map.get(
+                    (sample.container_name, sample.well_position)
+                ):
+                    well_position_to_sample_map[(sample.container_name, sample.well_position)] = []
+                well_position_to_sample_map[(sample.container_name, sample.well_position)].append(
+                    (case_index, sample_index)
+                )
+    return well_position_to_sample_map
```

Comment:
This method was chunky! Can we split it somehow or hide away the map in an object?

Here is one suggestion
```python
def validate_wells_contain_at_most_one_sample(order):
    well_position_map = WellPositionMap(order)
    errors = []

    for well_key in well_position_map.get_conflicting_wells():
        conflicting_indices = well_position_map.get_conflicting_indices(well_key)
        errors.extend(create_errors_for_conflicting_well(conflicting_indices))

    return errors


def create_errors_for_conflicting_well(conflicting_indices):
    return get_occupied_well_errors(conflicting_indices)


class WellPositionMap:
    def __init__(self, order):
        self.map = {}
        self._build_well_position_map(order)

    def _build_well_position_map(self, order):
        for case_index, case in order.enumerated_cases:
            self._add_samples_to_map(case_index, case)

    def _add_samples_to_map(self, case_index, case):
        for sample_index, sample in case.enumerated_samples:
            if self._is_sample_on_plate(sample):
                well_key = (sample.container_name, sample.well_position)
                self._add_sample_to_well_map(well_key, case_index, sample_index)

    def _add_sample_to_well_map(self, well_key, case_index, sample_index):
        if well_key not in self.map:
            self.map[well_key] = []
        self.map[well_key].append((case_index, sample_index))

    def _is_sample_on_plate(self, sample):
        return sample.container == ContainerEnum.plate

    def get_conflicting_wells(self):
        return [well_key for well_key, indices in self.map.items() if len(indices) > 1]

    def get_conflicting_indices(self, well_key):
        return self.map.get(well_key, [])[1:]

```

Reply:
I do agree it can be clearer! In this case, I have a hard time understanding the object though. "add_samples_to_map" does not give me any intel with regards to what actually happens and why. So all these help functions make me more confused 🫠 I'll think about it and maybe we can discuss what would be clearest?

Reply:
Yeah, I realize now that my example is messed up too haha

Reply:
Lets keep it and clean up in a separate PR 👍